core/src/main/java/org/apache/calcite/plan/RelOptPredicateList.java

```diff
@@ -236,14 +236,18 @@ public boolean isEffectivelyNotNull(RexNode e) {
       }
     }
     if (SqlKind.COMPARISON.contains(e.getKind())) {
-      // A comparison with a (non-null) literal, such as 'ref < 10', is not null if 'ref'
-      // is not null.
+      // A comparison with a (non-null) literal, such as 'ref < 10', or '10 < ref',
+      // is not null if 'ref' is not null.
       List<RexNode> operands = ((RexCall) e).getOperands();
       // We can have just one operand in case e.g. of a RexSubQuery with IN operator.
       if (operands.size() > 1 && operands.get(1) instanceof RexLiteral
           && !((RexLiteral) operands.get(1)).isNull()) {
         return isEffectivelyNotNull(operands.get(0));
       }
+      if (operands.size() > 1 && operands.get(0) instanceof RexLiteral
```

Comment:
@NobiGo this is to improve the `isEffectivelyNotNull` logic for `10 < ref` when pulled up predicates contain `ref is not null`, then `10 < ref`'s `isEffectivelyNotNull` will be `true`. This is useful to simplify the nested `RexNode`, like `is not true(10 < ref)`, and it will be simplified to `10 >= ref` which is more concise and useful for further simplification in this Jira.

Besides, another reasonable way is to split this part of code to another Jira, and it will be more clear why make these changes. @NobiGo What do you think?